File: solutions/fsharp/protein-translation/1/ProteinTranslation.fs
module ProteinTranslation

let proteinMap = Map[
    ("AUG", "Methionine"); 
    ("UUU", "Phenylalanine"); ("UUC", "Phenylalanine");
    ("UUA", "Leucine"); ("UUG", "Leucine");
    ("UCU", "Serine"); ("UCC", "Serine"); ("UCA", "Serine"); ("UCG", "Serine");
    ("UAU", "Tyrosine"); ("UAC", "Tyrosine"); 
    ("UGU", "Cysteine"); ("UGC", "Cysteine"); 
    ("UGG", "Tryptophan"); 
    ("UAA", "STOP"); ("UAG", "STOP"); ("UGA", "STOP")
    ]

let proteins rna = 
    let list = new List<string>()
    let mutable skip = 0
    let take = 3
    let mutable Break = false
    while(skip + take <= rna.Length || not Break) do
        let protein = string.Concat(rna.Skip(skip).Take(take))
        let translate = proteinMap[protein]
        if translate == "STOP" then
            Break <- true
        list.Add(translate)
        skip += take
    list.ToArray()

File: solutions/csharp/protein-translation/2/ProteinTranslation.cs
using System;
using System.Linq;
using System.Collections.Generic;

public static class ProteinTranslation
{
    private static readonly Dictionary<string, string> Protiens = new Dictionary<string, string>()
    {
        {"AUG", "Methionine"},
        {"UUU", "Phenylalanine"}, {"UUC", "Phenylalanine"},
        {"UUA", "Leucine"},{"UUG", "Leucine"},
        {"UCU", "Serine"},{"UCC", "Serine"},{"UCA", "Serine"},{"UCG", "Serine"},
        {"UAU", "Tyrosine"},{"UAC", "Tyrosine"},
        {"UGU", "Cysteine"},{"UGC", "Cysteine"},
        {"UGG", "Tryptophan"},
        {"UAA", "STOP"},{"UAG", "STOP"},{"UGA", "STOP"}
    };

    public static string[] Proteins(string strand)
    {
        var list = new List<string>();
        var skip = 0;
        const int take = 3;
        while (skip + take <= strand.Length)
        {
            var protien = string.Concat(strand.Skip(skip).Take(take));
            var translate = Protiens[protien];
            if (translate == "STOP") break;
            list.Add(translate);
            skip += take;
        }
        return list.ToArray();
    }
}